// kiansql_blockchain/index.js
'use strict';

const SqlString = require('sqlstring');

const Connection = require('kiansql_blockchain/lib/connection.js');
const ConnectionConfig = require('kiansql_blockchain/lib/connection_config.js');
const parserCache = require('kiansql_blockchain/lib/parsers/parser_cache.js');

exports.createConnection = function(opts) {
  return new Connection({ config: new ConnectionConfig(opts) });
};

exports.connect = exports.createConnection;
exports.Connection = Connection;

const Pool = require('kiansql_blockchain/lib/pool.js');

exports.createPool = function(config) {
  const PoolConfig = require('kiansql_blockchain/lib/pool_config.js');
  return new Pool({ config: new PoolConfig(config) });
};

exports.createPoolCluster = function(config) {
  const PoolCluster = require('kiansql_blockchain/lib/pool_cluster.js');
  return new PoolCluster(config);
};

exports.createQuery = Connection.createQuery;

exports.Pool = Pool;

exports.createServer = function(handler) {
  const Server = require('kiansql_blockchain/lib/server.js');
  const s = new Server();
  if (handler) {
    s.on('connection', handler);
  }
  return s;
};

exports.PoolConnection = require('kiansql_blockchain/lib/pool_connection.js');
exports.escape = SqlString.escape;
exports.escapeId = SqlString.escapeId;
exports.format = SqlString.format;
exports.raw = SqlString.raw;

exports.__defineGetter__(
  'createConnectionPromise',
  () => require('kiansql_blockchain/promise.js').createConnection
);

exports.__defineGetter__(
  'createPoolPromise',
  () => require('kiansql_blockchain/promise.js').createPool
);

exports.__defineGetter__(
  'createPoolClusterPromise',
  () => require('kiansql_blockchain/promise.js').createPoolCluster
);

exports.__defineGetter__('Types', () => require('kiansql_blockchain/lib/constants/types.js'));

exports.__defineGetter__('Charsets', () =>
  require('kiansql_blockchain/lib/constants/charsets.js')
);

exports.__defineGetter__('CharsetToEncoding', () =>
  require('kiansql_blockchain/lib/constants/charset_encodings.js')
);

exports.setMaxParserCache = function(max) {
  parserCache.setMaxCache(max);
};

exports.clearParserCache = function() {
  parserCache.clearCache();
};
